File: learn/src/main/java/makina/learn/neural/network/NetworkState.java
package makina.learn.neural.network;

import makina.math.matrix.Vector;
import makina.math.matrix.VectorType;
import makina.math.matrix.Vectors;

import java.util.HashMap;
import java.util.Map;

/**
 * // TODO: Add support for variables initialization method.
 *
 * @author Emmanouil Antonios Platanios
 */
class NetworkState {
    private final Map<Variable, Vector> variableValues = new HashMap<>();

    private final VariablesManager variablesManager;
    private final VectorType vectorType;

    NetworkState(VariablesManager variablesManager) {
        this(variablesManager, VectorType.DENSE);
    }

    NetworkState(VariablesManager variablesManager, VectorType vectorType) {
        this.variablesManager = variablesManager;
        this.vectorType = vectorType;
        for (Variable variable : variablesManager.variables())
            variableValues.put(variable, Vectors.build(variable.size, vectorType));
    }

    VectorType vectorType() {
        return vectorType;
    }

    void set(Variable variable, Vector value) {
        variableValues.put(variable, value);
    }

    void set(int id, Vector value) {
        variableValues.put(variablesManager.get(id), value);
    }

    void set(String name, Vector value) {
        variableValues.put(variablesManager.get(name), value);
    }

    Vector get(Variable variable) {
        return variableValues.getOrDefault(variable, null);
    }

    Vector get(int id) {
        return variableValues.getOrDefault(variablesManager.get(id), null);
    }

    Vector get(String name) {
        return variableValues.getOrDefault(variablesManager.get(name), null);
    }
}

File: learn/src/main/scala/makina/scala/learn/neural/network/NetworkState.scala
package makina.scala.learn.neural.network

import breeze.linalg.DenseVector

import scala.collection.mutable

/**
  * TODO: Add support for variables initialization method.
  *
  * @author Emmanouil Antonios Platanios
  */
class NetworkState private (private val variablesManager: VariablesManager) {
  private val variables = mutable.Map[Variable, DenseVector[Double]]()
  for (variable <- variablesManager.variables)
    variables += variable -> DenseVector.zeros[Double](variable.size)

  def set(variable: Variable, value: DenseVector[Double]) = variables += variable -> value
  def set(id: Int, value: DenseVector[Double]) = variables += variablesManager.get(id) -> value
  def set(name: String, value: DenseVector[Double]) = variables += variablesManager.get(name) -> value

  def get(variable: Variable) = variables(variable)
  def get(id: Int) = variables(variablesManager.get(id))
  def get(name: String) = variables(variablesManager.get(name))
}

object NetworkState {
  def apply(variablesManager: VariablesManager) = new NetworkState(variablesManager)
}
